sklearn/_loss/_loss.pyx.tp

```diff
@@ -359,6 +374,41 @@ cdef inline double_pair cgrad_hess_pinball_loss(
     return gh
 
 
+# Huber Loss
+cdef inline double closs_huber_loss(
+    double y_true,
+    double raw_prediction,
+    double delta,
+) noexcept nogil:
+    cdef double abserr = fabs(y_true - raw_prediction)
+    return (0.5 * abserr**2 if abserr <= delta
+            else delta * (abserr - 0.5 * delta))
+
+
+cdef inline double cgradient_huber_loss(
+    double y_true,
+    double raw_prediction,
+    double delta,
+) noexcept nogil:
+    cdef double res = raw_prediction - y_true
+    return res if fabs(res) <= delta else delta * (1 if res >=0 else -1)
+
+
+cdef inline double_pair cgrad_hess_huber_loss(
+    double y_true,
+    double raw_prediction,
+    double delta,
+) noexcept nogil:
+    cdef double_pair gh
+    gh.val2 = raw_prediction - y_true             # used as temporary
+    gh.val1 = (                                   # gradient
+        gh.val2 if fabs(gh.val2) <= delta
+        else delta * (1 if gh.val2 >=0 else -1)
+    )
+    gh.val2 = 1 if fabs(gh.val2) <= delta else 0  # hessian
```

Comment:
I find the nested `if-else` statements a little hard to parse. What do you think of the following?

```python
cdef double res = raw_prediction - y_true
if fabs(res) <= delta:
    gh.val1 = res
    gh.val2 = 1
else:
    gh.val1 = delta if res >= 0 else -delta
    gh.val2 = 0
```

Reply:
If the temporary is important, then:

```python
cdef double gh.val2 = raw_prediction - y_true
if fabs(gh.val2) <= delta:
    gh.val1 = res
    gh.val2 = 1
else:
    gh.val1 = delta if gh.val2 >= 0 else -delta
    gh.val2 = 0
```

Reply:
I was thinking about branchless code. If we don’t care, let’s do the simple if-else.